README.md
# node-virtualbox | [![dependencies Status](https://david-dm.org/ottomatica/node-virtualbox/status.svg)](https://david-dm.org/ottomatica/node-virtualbox)

This is a simple tool that helps provision basic VirtualBox virtual machines with sane defaults.

### Installation and Usage

Requires node >= 8.X

```
npm install node-virtualbox [--save] [-g]
```

Example run:

``` bash
node bin.js --provision --vmname "hello" --ip 172.168.0.55 --verbose
```

ssh into instance.
```
ssh -i config/resources/insecure_private_key -p 2002 -o StrictHostKeyChecking=no -o IdentitiesOnly=yes vagrant@127.0.0.1
```

### Default setup

The default VM will have 2 cpus and 1G memory. The default image is based on the latest [ubuntu/xenial64](https://cloud-images.ubuntu.com/xenial/current/) image. The VM has two NICs. The first nic uses NAT to forward incoming and outgoing traffic. The second nic is assigned a private host only network address. After creation, you can login with vagrant/vagrant or ssh with the default insecure_private_key located in `config/resources/`.

### Commands

`--list` return a list of vm names and uuids.

```
node bin.js --list
```

`--stop` Stop vm with save state.

```
node bin.js --stop --vmname <name>
```

`--delete` Unregister vm and delete all its contents.

```
node bin.js --delete --vmname <name>
```

`--info` Provide information about a vm. This will print out a json string with properties of the specified vm. 

For example, running `node .\bin.js --info --vmname "vm3"` will print the following, which can be used to retrieve properties such as the port that can be used for ssh access:

```
{ name: '"vm3"',
...
'Forwarding(0)': '"guestssh,tcp,,2002,,22"',
}
```

### Provision options

`--cpus` Set the number of cpus for VM. Default is 2 (or 1 for micro).

`--mem` Set the size of ram in MB(e.g., 512 or 1024). Default is 1024 (or 512 for micro).

`--ovf` Set the box to import when creating vm. If this is omitted, the latest ubuntu-xenial image is downloaded and used.

`--ssh_port` Set the local port used to forward ssh connections to vm. If this is omitted, then a freely available port between 2002 and 2999 is automatically assigned.

`--forward_ports` Set the port forwarding rules. Format: `"<guest_port>:<host_port>"` or `"<port>"`. `"<port>"` translates to `"<port>:<port>"`.

`--sync` Set a shared folder. Format: `"<host_folder>;<guest_folder>"`. You can provide multiple of these options.

`--add_ssh_key` Add public ssh key to ~/.ssh/authorized_keys of the vm.
```
node bin.js --provision --vmname "shared_folders_vm" --ip 172.16.1.45 --ssh_port 2095 --verbose --sync "C:\Users\chris;/chris" --sync "C:\Users\chris\projects;/projects" --add_ssh_key ~/.ssh/id_rsa.pub
```

### Micro VM
This is for booting micro kernels (custom initramfs inside of an iso)

Example micro VMs (iso):
``` bash
node bin.js --micro --vmname "micro" --attach_iso <Path>
```

If you don't specify path to an iso it will download and use an Alpine linux iso.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.35"></a>
## 0.1.35 (2019-02-13)


### Bug Fixes

* Sync folders should enable symlinks ([ebfe67e](https://github.com/ottomatica/node-virtualbox/commit/ebfe67e))



<a name="0.1.34"></a>
## 0.1.34 (2018-11-22)


### Bug Fixes

* Fix stop() to poweroff instead of saving state ([57b0cea](https://github.com/ottomatica/node-virtualbox/commit/57b0cea))

### Feature

* New `saveState()` to save state of VM (renamed old `stop()`) ([57b0cea](https://github.com/ottomatica/node-virtualbox/commit/57b0cea))


test/integration/micros.js
const child_process = require('child_process');
const os   = require('os');
const chai = require('chai');
const expect = chai.expect;
const path = require('path');

const util = require('../../lib/util');

describe('node-virtualbox should create micro vm with disk and shared folder, and stop and destroy it', function() {
    this.timeout(2000000);
    it('should create micro vm', function(done) {
        let testSharedPath = path.join(os.homedir(),'.baker');
        var child = child_process.exec(`node bin.js --micro --vmname "micro-vm" --ssh_port 2050 --verbose --disk --sync "${testSharedPath};/data"`, 
                                       {}, function(error, stdout, stderr) 
        {
            if( error ) console.log(stderr || stdout);
            expect(error).to.be.null;
            done();
        });
        child.stdout.pipe(process.stdout);
    });

    it('should have shared folder in micro vm', async function() 
    {
        let cmd = `ls /data`;
        let sshKeyPath = path.join('config','resources','baker_rsa');
        let buffer = await util.sshExec(cmd, {port: 2050, user: 'root', private_key: sshKeyPath}, 60000);
        expect(buffer).to.include('boxes');
    });

    it('should stop micro vm', function(done) {
        // echo value for prompt input for password.
        var child = child_process.exec(`node bin.js --stop --vmname "micro-vm"`, 
                                       {}, function(error, stdout, stderr) 
        {
            if( error ) console.log(stderr || stdout);
            expect(error).to.be.null;
            
            done();
        });
        child.stdout.pipe(process.stdout);
    });

    it('should destroy micro vm', function(done) {
        // echo value for prompt input for password.
        var child = child_process.exec(`node bin.js --delete --vmname "micro-vm"`, 
                                       {}, function(error, stdout, stderr) 
        {
            if( error ) console.log(stderr || stdout);
            expect(error).to.be.null;
            
            done();
        });
        child.stdout.pipe(process.stdout);
    });

});


test/integration/sanity-check.js
const child_process = require('child_process');
const chai = require('chai');
const expect = chai.expect;
const os = require('os');
const path = require('path');
const fs = require('fs-extra');

describe('node-virtualbox should create pingable vm, and stop and destroy it', function() {
    this.timeout(2000000);
    it('should create pingable vm', function(done) {
        // echo value for prompt input for password.
        var child = child_process.exec(`node bin.js --provision --vmname "sanity-check-vm" --ip 172.16.1.44 --port 2092 --verbose`, 
                                       {}, function(error, stdout, stderr) 
        {
            if( error ) console.log(stderr || stdout);
            expect(error).to.be.null;

            let cmd = process.platform === 'win32' ? `ping 172.16.1.44 -n 5` : `ping 172.16.1.44 -c 5`
            let output = child_process.execSync(cmd).toString();
            console.log(output);
            expect(output).to.include('time=');

            done();
        });
        child.stdout.pipe(process.stdout);
    });

    it('should stop vm', function(done) {
        // echo value for prompt input for password.
        var child = child_process.exec(`node bin.js --stop --vmname "sanity-check-vm"`, 
                                       {}, function(error, stdout, stderr) 
        {
            if( error ) console.log(stderr || stdout);
            expect(error).to.be.null;
            
            done();
        });
        child.stdout.pipe(process.stdout);
    });

    it('should destroy vm', function(done) {
        // echo value for prompt input for password.
        var child = child_process.exec(`node bin.js --delete --vmname "sanity-check-vm"`, 
                                       {}, function(error, stdout, stderr) 
        {
            if( error ) console.log(stderr || stdout);
            expect(error).to.be.null;
            
            done();
        });
        child.stdout.pipe(process.stdout);
    });

});


test/integration/shared-folders.js
const child_process = require('child_process');
const chai = require('chai');
const expect = chai.expect;
const path = require('path');

const util = require('../../lib/util');

describe('node-virtualbox should create vm with shared folder, and stop and destroy it', function() {
    this.timeout(2000000);
    it('should create vm', function(done) {
        let testSharedPath = __dirname;
        var child = child_process.exec(`node bin.js --provision --vmname "shared-folders-vm" --ip 172.16.1.47 --port 2097 --verbose --sync "${testSharedPath};/testShare"`, 
                                       {}, function(error, stdout, stderr) 
        {
            if( error ) console.log(stderr || stdout);
            expect(error).to.be.null;
            done();
        });
        child.stdout.pipe(process.stdout);
    });

    it('should have shared folder in vm', async function() 
    {
        let cmd = `ls /testShare`;
        let sshKeyPath = path.join('config','resources','insecure_private_key');
        let buffer = await util.sshExec(cmd, {port: 2097, user: 'vagrant', private_key: sshKeyPath}, 60000);
        expect(buffer).to.include('shared-folders.js');
    });

    it('should stop vm', function(done) {
        // echo value for prompt input for password.
        var child = child_process.exec(`node bin.js --stop --vmname "shared-folders-vm"`, 
                                       {}, function(error, stdout, stderr) 
        {
            if( error ) console.log(stderr || stdout);
            expect(error).to.be.null;
            
            done();
        });
        child.stdout.pipe(process.stdout);
    });

    it('should destroy vm', function(done) {
        // echo value for prompt input for password.
        var child = child_process.exec(`node bin.js --delete --vmname "shared-folders-vm"`, 
                                       {}, function(error, stdout, stderr) 
        {
            if( error ) console.log(stderr || stdout);
            expect(error).to.be.null;
            
            done();
        });
        child.stdout.pipe(process.stdout);
    });

});


bin.js
#!/usr/bin/env node

/**
 * Run virtualbox as program.
 *
 * @module virtualbox/bin
 */

// Modules
const yargs     = require('yargs');
const _         = require('lodash');
const virtualbox = require('./index');

// Dockerize
(async () => {

    try {

        // Get args and validate
        let argv = yargs.boolean('dry-run').boolean('verbose').argv;
        let args = argv._;
        if (args.length > 1) {
            throw new Error('Usage: virtualbox [--vmname=] [--verbose] ...');
        }

        // common options
        let vmname = argv.vmname;
        let verbose = argv.verbose;

        // general commands
        let list = argv.list;
        let start = argv.start;
        let check = argv.check;
        let deleteCmd = argv.delete;
        let stopCmd   = argv.stop;
        let infoCmd   = argv.info;

        // provision related arguments
        let provision = argv.provision;
        let ovf = argv.ovf;
        let attach_iso = argv.attach_iso;
        let micro = argv.micro;
        let disk = argv.disk;
        let ip = argv.ip;
        let ssh_port = argv.ssh_port;
        let forward_ports = argv.forward_ports;
        let cpus = argv.cpus;
        let mem  = argv.mem;
        let add_ssh_key = argv.add_ssh_key;
        let syncs = [];
        if( argv.sync)
        {
            syncs = _.isArray(argv.sync) ? argv.sync : [ argv.sync ];
        }

        // Update running vm arguments
        let exposePort = argv.exposePort;

        // If a dry run, enable logging
        if (argv.dryRun) {
        }

        // Provision 
        await virtualbox(
            {
                vmname, ovf, verbose, list, start, check, provision, ip, ssh_port, deleteCmd, stopCmd, infoCmd, syncs, attach_iso, micro,
                mem, cpus, add_ssh_key, forward_ports, disk, exposePort
            }
        );

    }
    catch(e) {
        console.log(e);
        process.exit(1);
    }

})();

index.js
/**
 * @module virtualbox
 */

const path          = require('path');
const fs            = require('fs');

const download      = require('download');
const tar           = require('tar');
const md5File       = require('md5-file/promise')
const ProgressBar = require('progress');

const util          = require('./lib/util');
const VBoxProvider  = require('./lib/VBoxProvider');

module.exports = async function (options = {}) {
    let provider = new VBoxProvider();

    if( !options.ssh_port && (options.provision || options.micro) )
    {
        options.ssh_port = await util.findAvailablePort(provider, options.verbose);
    }

    if( !options.cpus && (options.provision || options.micro) )
    {
        options.cpus = options.provision ? 2: 1;
    }

    if( !options.mem && (options.provision || options.micro) )
    {
        options.mem = options.provision ? 1024: 512;
    }

    if(options.micro) {
        try {
            let iso = 'https://github.com/ottomatica/baker-release/releases/download/latest-dev/alpine.iso';
            const boxesPath = path.join(require('os').userInfo().homedir, '.baker', 'boxes');
            const isoPath = options.attach_iso || path.join(boxesPath, 'alpine.iso');
            if (!(await fs.existsSync(isoPath)) || (await md5File(isoPath)) != '851e2b2b34e31b67aa0758d25666e8e5') {
                await download(iso, boxesPath);
            }
    
            await provider.micro(options.vmname, options.cpus, options.mem, isoPath, options.ssh_port, path.join(__dirname,'config/resources/baker_rsa'), options.syncs, options.disk, options.verbose);
       } catch (error) {
            console.error('=> exec error:', error);
        }
    }

    if(options.provision) {

        if( !options.ovf )
        {
            const boxesPath = path.join(require('os').userInfo().homedir, '.baker', 'boxes');
            const unpackPath = path.join(boxesPath, 'ubuntu-xenial');

            util.mkDirByPathSync(boxesPath);
            util.mkDirByPathSync(unpackPath);

            // download files if not available locally
            if (!(await fs.existsSync(path.join(unpackPath, 'box.ovf')))) {
                console.log("no --ovf specified, downloading latest ubuntu box!");
                const bar = new ProgressBar('[:bar] :percent :etas', {
                    complete: '=',
                    incomplete: ' ',
                    width: 20,
                    total: 0
                });

                await download('http://cloud-images.ubuntu.com/xenial/current/xenial-server-cloudimg-amd64-vagrant.box', boxesPath)
                      .on('response', res => {
                        // console.log(`Size: ${res.headers['content-length']}`);
                        bar.total = res.headers['content-length'];
                        res.on('data', data => bar.tick(data.length));
                      })
                      .then(() => console.log('downloaded!'));
                await tar.x(  // or tar.extract(
                    {
                      file: path.join(boxesPath, 'xenial-server-cloudimg-amd64-vagrant.box'),
                      C: unpackPath
                    }
                );
                // Remove box
                fs.unlinkSync(path.join(boxesPath, 'xenial-server-cloudimg-amd64-vagrant.box'));
            }
            options.ovf = path.join(unpackPath, 'box.ovf');
        }

        try {
            await provider.check(options);
            await provider.provision(options.vmname, options.cpus, options.mem, options.ovf, options.attach_iso, options.verbose);
            await provider.customize(options.vmname, options.ip, options.ssh_port, options.forward_ports, options.syncs, options.verbose);
            await provider.start(options.vmname, options.verbose);
            await provider.postSetup(options.vmname, options.ip, options.ssh_port, path.join(__dirname,'config/resources/insecure_private_key'), options.add_ssh_key, options.syncs, options.verbose);
        } catch (error) {
            console.error('=> exec error:', error);
        }
    }
    
    if(options.list)
        console.log(await provider.list());

    if(options.deleteCmd)
    {
        if( !options.vmname )
        {
            console.error("Please provide --vmname <name> with --delete");
            process.exit(1);
        }        
        console.log(await provider.delete(options.vmname));
    }

    if(options.stopCmd)
    {
        if( !options.vmname )
        {
            console.error("Please provide --vmname <name> with --stop");
            process.exit(1);
        }        
        console.log(await provider.stop(options.vmname));
    }

    if(options.infoCmd){

        if( !options.vmname )
        {
            console.error("Please provide --vmname <name> with --info");
            process.exit(1);
        }
        console.log(await provider.info(options.vmname)); 
    }

    if(options.start)
        await provider.start(options.vmname, options.verbose);

    if(options.exposePort )
    {
        if( !options.vmname )
        {
            console.error("Please provide --vmname <name> with --exposePort");
            process.exit(1);
        }

        console.log(await provider.expose(options.vmname, options.exposePort, options.verbose));
    }
};


lib/VBoxProvider.js
/**
 * @module lib/VBoxProvider
 */

const execute = require('./commands/execute');
const execSync = require('child_process').execSync;
const exec = require('child_process').exec;
const { promisify } = require('util');
const execAsync = promisify(exec);
const mustache = require('mustache');
const os = require('os');
const fs = require('fs-extra');
const path = require('path');
const ipUtil = require('ip');
const util = require('../lib/util');
const VBexe = process.platform === 'win32' ? '"C:\\Program Files\\Oracle\\VirtualBox\\VBoxManage.exe"' : 'VBoxManage';

const isPortAvailable = require('is-port-available');

class VBoxProvider {

    /**
     * Get default run command.
     *
     * @param   {String} name Name of virtual machine.
     * @param   {String} ovf  Path of image to clone/import.
     * @returns {Object}      Promise.
     */
    async provision(name, cpus, mem, ovf, iso, verbose) {
        await execute("import", `"${ovf}" --vsys 0 --vmname ${name}`, verbose);
        await execute("modifyvm", `"${name}" --memory ${mem} --cpus ${cpus}`, verbose);
        if(iso)
            await execute("storageattach", `${name} --storagectl IDE --port 0 --device 0 --type dvddrive --medium "${iso}"`)
    }

    _calculateGateway(ip, mask='255.255.255.0'){
        let networkAddress = ipUtil.mask(ip, mask);
        return ipUtil.cidrSubnet(networkAddress + '/26').firstAddress;
    }

    async micro(name, cpus, mem, iso, ssh_port, sshKeyPath, syncs, disk, verbose) {
        await execute("createvm", `--name "${name}" --register`, verbose);
        await execute("modifyvm", `"${name}" --memory ${mem} --cpus ${cpus}`, verbose);
        await execute("storagectl", `"${name}" --name IDE --add ide`, verbose);
        await execute("storageattach", `${name} --storagectl IDE --port 0 --device 0 --type dvddrive --medium "${iso}"`, verbose);

        await execute("modifyvm", `${name}  --uart1 0x3f8 4 --uartmode1 disconnected`, verbose);
        
        // NIC1 =======
        await execute("modifyvm", `${name} --nic1 nat`, verbose);
        await execute("modifyvm", `${name} --nictype1 virtio`, verbose);
        
        // port forwarding
        await execute("modifyvm", `${name} --natpf1 "guestssh,tcp,,${ssh_port},,22"`, verbose);
        
        // Disks
        if( disk )
        {
            let diskPath = path.join(os.homedir(),'.baker','boxes',`${name}-disk.vdi`);
            let diskSize = 32768;
            if( !fs.existsSync(diskPath) )
            {
                //let rawDisk = "\\\\.\\PhysicalDrive0";
                //await execute("internalcommands", `createrawvmdk -filename ${diskPath} -rawdisk "${rawDisk}"`, verbose);
                // --variant Fixed
                await execute("createmedium", `disk --format VDI --filename "${diskPath}" --size ${diskSize} `, verbose);
                //await execute("modifymedium", `${diskPath} --type writethrough`, verbose);
            }

            await execute("storagectl", `"${name}" --name "SATA" --add sata`, verbose);
            await execute("storageattach", `${name} --storagectl "SATA" --port 2 --device 0 --type hdd --medium "${diskPath}"`, verbose);
        }

        // syncs
        if( syncs.length > 0 )
        {
            let count = 0;
            for( var sync of syncs )
            {
                let host = sync.split(';')[0];
                let guest = sync.split(';')[1];
                await execute("sharedfolder", `add ${name} --name "vbox-share-${count}" --hostpath "${host}" `, verbose);
                count++;
            }
        }

        await this.start(name, verbose);

        // post setup
        if( disk )
        {
            // format new disk to be ext4
            let sshConfig = {port: ssh_port, user: 'root', private_key: sshKeyPath};
            //let cmd = 'echo -e "o\nn\np\n1\n\n\nw" | fdisk /dev/sda && /sbin/mkfs.ext4 /dev/sda1';
            let cmd = `/sbin/mkfs.ext4 /dev/sda; mount -t ext4 /dev/sda /mnt/disk`;
            console.log('Formating virtual drive');

            await util.sshExec(cmd, sshConfig, 60000, verbose).catch( e => {console.log(e)});;
        }
    }

    async customize(name, ip, ssh_port, forward_ports=[], syncs, verbose) {
        // modifyvm ${VM} --uart1 0x3f8 4 --uartmode1 disconnected
        await execute("modifyvm", `${name}  --uart1 0x3f8 4 --uartmode1 disconnected`, verbose);
        
        // syncs
        if( syncs.length > 0 )
        {
            let count = 0;
            for( var sync of syncs )
            {
                let host = sync.split(';')[0];
                let guest = sync.split(';')[1];
                await execute("sharedfolder", `add ${name} --name "${name}-${count}" --hostpath "${host}" `, verbose);
                await execute(`setextradata`, `${name} VBoxInternal2/SharedFoldersEnableSymlinksCreate/${name}-0 1`, verbose);
                count++;
            }
        }
        
        // NIC1 =======
        await execute("modifyvm", `${name} --nic1 nat`, verbose);
        await execute("modifyvm", `${name} --nictype1 virtio`, verbose);
        
        if( ip )
        {
            // NIC2 =======
            let VBOXNET = null;
            // check if any adapters with this ip :
            let gateway = this._calculateGateway(ip);
            let networks = (await this.hostonlyifs()).filter(e => e.IPAddress === gateway);
            if (networks.length > 0 )
            {
                VBOXNET = networks[0].Name;
                console.log(`Using ${gateway} in ${VBOXNET}`);
            }
            else 
            {
                let stdout = (await execAsync(`${VBexe} hostonlyif create`)).stdout;
                VBOXNET = stdout.substr(stdout.indexOf(`'`) + 1, stdout.lastIndexOf(`'`) - stdout.indexOf(`'`) - 1);
                console.log('created adapter:', VBOXNET);
            }
            
            await execute("hostonlyif", `ipconfig "${VBOXNET}" --ip ${gateway}`, verbose);
            
            await execute("modifyvm", `${name} --hostonlyadapter2 "${VBOXNET}"`, verbose);
            await execute("modifyvm", `${name} --nic2 hostonly`, verbose);
            await execute("modifyvm", `${name} --nictype2 virtio`, verbose);
        }
        
        // port forwarding for ssh
        await execute("modifyvm", `${name} --natpf1 "guestssh,tcp,,${ssh_port},,22"`, verbose);
        
        // port forwarding
        forward_ports.forEach(async function (port) {
            let splitPort = String(port).split(':');
            let guestPort = splitPort[0];
            let hostPort = splitPort[1] || splitPort[0];
            await execute("modifyvm", `${name} --natpf1 "${port},tcp,,${hostPort},,${guestPort}"`, verbose);
        });
    }
    
    async start(name, verbose) {
        // For unlock any session.
        await execute("startvm", `${name} --type emergencystop`, verbose).catch(e => e);
        // Real start.
        await execute("startvm", `${name} --type headless`, verbose);
    }
    
    async waitForConnection(sshInfo, tries=0)
    {
        try {
            let cmd = `echo "waiting to start"`;
            await util.sshExec(cmd, sshInfo, 60000);
        } catch (error) {
            if( tries > 5 )
            {
                throw `Timed out.`;
            }
            this.waitForConnection(sshInfo,tries++);
        }
    }
    
    // setting /etc/network/interfaces
    async postSetup(vmname, ip, port, sshKeyPath, newSSHKeyPath, syncs, verbose) {
        let interfacesPath = path.resolve(__dirname, '../config/interfaces.mustache');
        
        // render and create interfaces in /tmp/interfaces
        let tmpFile = path.join(os.tmpdir(),'interfaces');
        let interfaces = mustache.render((await fs.readFile(interfacesPath)).toString(), {ip});
        await fs.writeFile(tmpFile, interfaces);
        
        await this.waitForConnection({port, user: 'vagrant', private_key: sshKeyPath});
        
        if (ip )
        {
            // cp /tmp/interfaces vm:/tmp/interfaces
            try {
                await util.scp(tmpFile, '/tmp/interfaces', {port, user: 'vagrant', private_key: sshKeyPath});
            } catch (error) {
                throw `failed to generate interfaces configuration, ${error}`;
            }
            
            // mv vm:/tmp/interfaces /etc/network/interfaces, 
            // sudo systemctl restart networking
            // ifdown ens0p8; ifup ens0p8
            try {
                let cmd = `sudo cp /tmp/interfaces /etc/network/interfaces && sudo systemctl restart networking && sudo ifdown enp0s8 && sudo ifup enp0s8`;
                await util.sshExec(cmd, {port, user: 'vagrant', private_key: sshKeyPath}, 60000, verbose);
            } catch (error) {
                throw `failed to copy interfaces configuration, ${error}`;
            }
        }
        

        // adding new ssh-key
        if(newSSHKeyPath){
            try {
                let newSSHKey = (await fs.readFile(newSSHKeyPath)).toString();
                let cmd = `echo "${newSSHKey}" >> ~/.ssh/authorized_keys`;
                await util.sshExec(cmd, {port, user: 'vagrant', private_key: sshKeyPath}, 60000, verbose);
            } catch (error) {
                throw `failed to add new ssh key, ${error}`;
            }
        }

        await this.setupSyncFoldersOnGuest(vmname, syncs, port, sshKeyPath, verbose);
       
    }

    async setupSyncFoldersOnGuest(vmname, syncs, port, sshKeyPath, verbose)
    {
       // Handle sync folders
       if( syncs.length > 0 )
       {
           // Add vboxsf to modules so we can enable shared folders; ensure our user is in vboxsf group
           try {
               let LINE =  "vboxsf"; let FILE= '/etc/modules';
               let cmd = `(grep -qF -- "${LINE}" "${FILE}" || echo "${LINE}" | sudo tee -a "${FILE}"); sudo usermod -a -G vboxsf vagrant`;
               await util.sshExec(cmd, {port, user: 'vagrant', private_key: sshKeyPath}, 60000, verbose);
           } catch (error) {
               throw `failed to setup shared folders, ${error}`;
           }

           // Add mount to /etc/fstab for every shared folder
           let count = 0;
           for( var sync of syncs )
           {
               let host = sync.split(';')[0];
               let guest = sync.split(';')[1];

               try {
                   let LINE=`${vmname}-${count}    ${guest}   vboxsf  uid=1000,gid=1000   0   0`; let FILE=`/etc/fstab`; 
                   let cmd = `sudo mkdir -p ${guest}; grep -qF -- "${LINE}" "${FILE}" || echo "${LINE}" | sudo tee -a "${FILE}"`;
                   await util.sshExec(cmd, {port, user: 'vagrant', private_key: sshKeyPath}, 60000, verbose);
               } catch (error) {
                   throw `failed to add fstab entry for shared folder, ${error}`;
               }
               count++;
           }

           // Reload fstab
           try {
               let cmd = `sudo mount -a`;
               await util.sshExec(cmd, {port, user: 'vagrant', private_key: sshKeyPath}, 60000, verbose);
           } catch (error) {
               throw `failed to setup shared folders, ${error}`;
           }
           
       }        
    }

    async list() {
        return new Promise(function (resolve, reject) {   
            exec(`${VBexe} list vms`, (error, stdout, stderr) => {
                if(error || stderr) {
                    console.error(`exec error: vboxmanage list`);
                    console.error(`=> ${error}, ${stderr}`);
                    reject(error);
                }

                let list = [];
                let lines = stdout.split('\n');
                for (let i = 0; i < lines.length-1; i++) {
                    let lineSplit = lines[i].split(' ');
                    let name= lineSplit[0].replace(/"/g, '');
                    let id = lineSplit[1].replace(/{|}/g, '');
                    list.push({name: name, id: id});
                }
                resolve(list);
            })
        });
    }

    async hostonlyifs(){
        return new Promise(function (resolve, reject) {   
            exec(`${VBexe} list hostonlyifs`, (error, stdout, stderr) => {
                if(error || stderr) {
                    console.error(`exec error: vboxmanage list`);
                    console.error(`=> ${error}, ${stderr}`);
                    reject(error);
                }

                let hostonlyifs = [];
                stdout.split(/\r?\n\r?\n/).forEach(adapters => {
                    if(adapters.length > 0) {
                        let adapter = {};
                        adapters.split('\n').forEach(line => {
                            if(line.length > 0) {
                                let splitIdx = line.indexOf(':');
                                adapter[line.substr(0, splitIdx).trim()] = line.substr(splitIdx+1).trim();
                            }
                        })
                        hostonlyifs.push(adapter);
                    }
                })

                resolve(hostonlyifs);
            })
        });
    }

    async check(options) {
        let ovf =options.ovf;
        let name=options.vmname;
        let syncs=options.syncs;
        return new Promise(async function (resolve, reject) {
            // check if box already exists
            if(!(await fs.exists(path.resolve(ovf)))) {
                reject(`File not found ${ovf}`);
            }

            // Progress state: VBOX_E_FILE_ERROR
            // VBoxManage: error: Appliance import failed
            // VBoxManage: error: Machine settings file '/Users/cjparnin/VirtualBox VMs/hello/hello.vbox' already exists
            let MachineSettingPath = path.join(require('os').userInfo().homedir, `VirtualBox\ VMs/${name}/${name}.vbox`);
            if (await fs.exists(MachineSettingPath)) {
                reject(`Machine setting file ${MachineSettingPath} already exists.`);
            }

            // Verify correct format and existence of sync folders.
            for( var sync of syncs )
            {
                let atoms = sync.split(';');
                if( atoms.length !=2 )
                {
                    reject(`Invalid sync folder format. Please use this format: "<host_folder>;<guest_folder>".`);
                }
                let host = atoms[0];
                let guest = atoms[1];
                if (!await fs.exists(host)) {
                    reject(`The path ${host} does not exist on your host machine. Cannot create shared folder.`);
                }
            }

            // Verify new ssh key exists
            if (options.add_ssh_key) {
                const sshKeyExists = await fs.pathExists(options.add_ssh_key);
                if(!sshKeyExists)
                    reject(`No such file or directory: ${options.add_ssh_key}`);
            }

            resolve();
        })
    }

    async saveState(name) {
        return new Promise(function (resolve, reject) {   
            exec(`${VBexe} controlvm ${name} savestate`, (error, stdout, stderr) => {
                if(error && stderr.indexOf('VBOX_E_OBJECT_NOT_FOUND') == -1) {
                    console.error(`exec error: stop`);
                    console.error(`=> ${error}, ${stderr}`);
                    reject(error);
                }
                resolve("");
            });
        });
    }

    async stop(name) {
        return new Promise(function (resolve, reject) {   
            exec(`${VBexe} controlvm ${name} poweroff soft`, (error, stdout, stderr) => {
                if(error && stderr.indexOf('VBOX_E_OBJECT_NOT_FOUND') == -1) {
                    console.error(`exec error: stop`);
                    console.error(`=> ${error}, ${stderr}`);
                    reject(error);
                }
                resolve("");
            });
        });
    }

    async delete(name) {
        return new Promise(function (resolve, reject) {   
            exec(`${VBexe} unregistervm ${name} --delete`, (error, stdout, stderr) => {
                if(error && stderr.indexOf('VBOX_E_OBJECT_NOT_FOUND') == -1) {
                    console.error(`=> ${error}, ${stderr}`);
                    reject(error);
                }
                resolve("");
            });
        });
    }

    async info(vmname) {
        return new Promise(function (resolve, reject) {   
            exec(`${VBexe} showvminfo ${vmname} --machinereadable`, (error, stdout, stderr) => {
                if(error && stderr.indexOf('VBOX_E_OBJECT_NOT_FOUND') != -1) {
                    resolve({VMState:'not_found'});
                }
                else if( error )
                {
                    console.error(`=> ${error}, ${stderr}`);
                    reject(error);
                }
                else
                {
                    let properties = {};
                    let lines = stdout.split('\n');
                    for (let i = 0; i < lines.length-1; i++) {
                        let lineSplit = lines[i].split('=');
                        let name= lineSplit[0].trim();
                        let id = lineSplit[1].trim();
                        properties[name]=id;
                    }
                    resolve(properties);
                }
            });
        });
    }

    async getState(name) {
        let vmInfo = await this.info(name);
        return vmInfo.VMState.replace(/"/g,'');
    }

    async expose(name, port, verbose) {

        let hostPort=port, guestPort=port;
        if( port.indexOf(':') > 0 )
        {
            [hostPort,guestPort] = port.split(':');
        }

        var status = await isPortAvailable(hostPort);
        if(!status) 
        {
            throw new Error(`The port ${hostPort} is not available for use!`)
        }

        try
        {
            if( await this.getState(name) == "running" )
            {
                await execute("controlvm", `${name} natpf1 "${hostPort},tcp,,${hostPort},,${guestPort}"`, verbose);
            }
            else
            {
                await execute("modifyvm", `${name} --natpf1 "${hostPort},tcp,,${hostPort},,${guestPort}"`, verbose);
            }
        }
        catch(err)
        {
            if( err.message.indexOf("name already exists") == -1 )
            {
                throw new Error(err);
            }
        }
        return `Added exposed port ${hostPort} => ${guestPort} on ${name}`;
    }

}

// Export
module.exports = VBoxProvider;